File: 0224-basic-calculator/0224-basic-calculator.cpp
// @Author: KING-SEN
// Programming Language Used: C++

class Solution {
public:
    int calculate(string s) {
        int sum = 0, sign = 1;
        stack<int> stack;
        
        for(int i = 0; i < s.size(); i++){
            // Empty space
            if(s[i] == ' ')
                continue;
            // For number
            else if(isdigit(s[i])){
                int res = 0;
                while(i < s.size() && isdigit(s[i])){
                    res = res * 10 + (s[i] - '0');
                    i++;
                }
                i--;
                sum += res * sign;
            }
            // For +
            else if(s[i] == '+'){
                sign = 1;
            }
            // For -
            else if(s[i] == '-'){
                sign = -1;
            }
            else if(s[i] == '('){
                stack.push(sum);
                stack.push(sign);
                sum = 0;
                sign = 1;
            }
            // For )
            else{
                sum = stack.top() * sum;
                stack.pop();
                sum += stack.top();
                stack.pop();
            }
        }
        
        return sum;
    }
};

// Time Complexity: O(N), Here N = length of the string 
// Space Complexity: O(N)

File: 0224-basic-calculator/0224-basic-calculator.py
class Solution:
    def calculate(self, s: str) -> int:
        result, sign = 0, 1
        stack = []
        i = 0
        while i < len(s):
            if s[i] == " ": # For empty space
                i += 1
                continue
            elif s[i].isdigit(): # For digit
                res = 0
                while i < len(s) and s[i].isdigit():
                    res = res * 10 + int(s[i])
                    i += 1
                i -= 1
                result += res * sign
            elif s[i] == "+":
                sign = 1
            elif s[i] == '-':
                sign = -1
            elif s[i] == '(':
                stack.append(result)
                stack.append(sign)
                result, sign = 0, 1
            else:
                result = stack.pop() * result
                result += stack.pop()
            i += 1
        return result
            
                